enhancements/network/bgp-ovn-kubernetes.md

```diff
@@ -0,0 +1,689 @@
+* Increases support complexity due to integration with the a user's provider network.
+* Increases network control plane load on nodes, as they have multiple peer connections and are having to maintain
+a dynamic routing protocol.
+
+### Implementation Details/Notes/Constraints
+
+#### BGP Peering
+
+FRR-K8S shall support both internal BGP (iBGP) and external BGP (eBGP). With iBGP, a full mesh topology is required. This
+can be cumbersome and have issues with scaling in large clusters. In the future, FRR-K8S shall also support designating one
+or more nodes as route reflectors (https://issues.redhat.com/browse/CNF-10719). With route reflectors, nodes only have
+to connect to the route reflectors and do not have to maintain a full mesh with all other nodes in the cluster. Therefore,
+support for iBGP in this enhancement will be restricted to full mesh only.
+
+#### FRR-K8S Integration
+
+As previously mentioned frr-k8s will be used in order to deploy and manage FRR configuration. Support will be added
+where necessary for gaps in frr-k8s API, but in the meantime frr-k8s supports also issuing raw FRR configuration. Every
+attempt should be made to leverage the FRR-K8S API, and raw configuration should only be used as a last resort. A user
+will need the ability to:
+
+* Define BGP peers, ASN, authentication, along with other common BGP configurations like communities, route filtering, etc.
+* Configure BFD
+
+FRR-K8S will serve as the main API for user configuration of BGP via FRR-K8S CRs. OVN-Kubernetes will watch for these
+CRs, and once created, OVN-Kubernetes will start monitoring the kernel routing table via netlink for routes installed
+by FRR. When FRR is using Zebra, the routes can be identified via a proto "bgp" label:
+
+```
+123.123.123.0/24 nhid 47 via 172.18.0.5 dev breth0 proto bgp metric 20
+```
+
+When these routes are seen on the node, ovnkube-controller will update the routes in NBDB for any GW routers on the
+node, filtering the route for the GW router based on if the interface on the route matches the interface in OVS. This
+will happen irrespective of gateway mode, as some features in local gateway mode rely on the GRs for sending traffic
+(Egress IP, MEG). Note, this configuration will only happen on nodes selected by the FRR-K8S CRs.
+
+#### API Changes
+
+##### Route Advertisements
+
+When OVN-Kubernetes detects that a FRR-K8S CR has been created for BGP peering, OVN-Kubernetes will by default
+advertise the pod subnet for each node, via creating an additive, per node FRR-K8S CR that is managed by OVN-Kubernetes.
+OVN-Kubernetes CRD:
+
+```yaml
+apiVersion: k8s.ovn.org/v1
+kind: RouteAdvertisements
+metadata:
+  name: default
+spec:
+    # networkSelector:
+    # nodeSelector:
+    # frrConfigurationSelector:
+    targetVRF: default
+    advertisements:
+      podNetwork: true
+      egressIP: true
+      clusterIPs: true
+      externalIPs: true
```

Comment:
I am not sure we should export cluster IPs and external IPs just because we can. It is inherent in their definitions that these IPs should not be managed outside of cluster boundaries by us:
- cluster IPs are just defined to be used from within the cluster
- external IPs are defined as externally managed. Our only responsibility towards them is once and if they reach the cluster, to forward to the appropriate backend. So it is not our responsibility per definition to facilitate the routing towards the cluster.
I would rather take a conservative approach and listen to specific use cases before we support this king of thing.
@fedepaol  thoughts?

Reply:
Agree. 

I also discussed this with @trozet today:

- There's no point in advertising clusterips (to make the networking of the cluster work) as the lb is always going to be resolved locally to the pod ips and then bgp will help routing.
- External ips would become just a simpler (and possibly less configurable) version of loadbalancers

Reply:
we all agree. I will remove these fields in the enhancement